### PR TITLE
DQM modules to monitor Pixel RecHits for heterogenous workflows 

### PR DESCRIPTION
**Pull request by sroychow**
#### PR description:

This PR introduces Rechi monitoring modules for pixels for CPU and GPU workflows. There is a base class where the histo booking, filling are done.  Related [talk](https://indico.cern.ch/event/1134995/#13-gpu-cpu-dqm-modules-for-tra)

#### PR validation:
Checks done with 11634.501, 11634.506

### CODE REVIEW

**Comment by tsusa** [comment]
please test

**Comment by sroychow** [comment]
enable gpu

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc
Comment (mmusich) [inline DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc:63]: this you can take from the ES.
See e.g. at:

https://github.com/cms-sw/cmssw/blob/772910aeac89026866ae4c7a7246e33f49a8e300/CalibTracker/SiPixelLorentzAngle/src/SiPixelLorentzAnglePCLHarvester.cc#L139-L146

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc
Comment (mmusich) [inline DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc:77]: see above.

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc
Comment (mmusich) [inline DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc:89]: instead of having a large set of arguments passed, wouldn't it be more elegant to store them in a POD `struct` or something to that effect?

**Review by mmusich — COMMENTED** [review]
some quick preliminary comments.

**Comment by tsusa** [comment]
urgent 
needed for GPU/CPU comparison in 123X

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc
Comment (perrotta) [inline DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc:26]: Why not in the class initialization list (and made constant)?

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc
Comment (perrotta) [inline DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc:27]: Why not in the class initialization list (and made constant)?

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc
Comment (mmusich) [inline DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc:11]: what is this needed for?

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoACPU.cc
Comment (mmusich) [inline DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoACPU.cc:11]: what is this needed for?

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc
Comment (arossi83) [reply DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc:77]: also the number of sides is accessible from geometry? I cannot find the proper method. (the other hard coded numbers will be replaced as you suggested)

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoACPU.cc
Comment (mmusich) [inline DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoACPU.cc:24]: could be `const` (if then put into the initializer list)

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoAGPU.cc
Comment (mmusich) [inline DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoAGPU.cc:29]:
```suggestion
  const edm::EDGetTokenT<cms::cuda::Product<TrackingRecHit2DGPU>> tokenHitsGPU_;
  const edm::EDGetTokenT<SiPixelClusterCollectionNew> clusterToken_;
```

if they're put into the initializer's list at line 40-41.

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoAGPU.cc
Comment (mmusich) [inline DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoAGPU.cc:13]: this comment doesn't seem to be related?

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc
Comment (sroychow) [reply DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc:11]: removed

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoAGPU.cc
Comment (mmusich) [reply DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoAGPU.cc:29]: also the convention for the member data is inconsistent (either use leading `m_` or trailing `_`)

**Review by mmusich — COMMENTED** [review]
this other set is minor and I guess it could go into a follow-up.

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc
Comment (fwyzard) [inline DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc:27]: I would prefer _not_ to have a parameter without any effect, because setting it to `True` or `False` will be misleading :-(

Comment (fwyzard) [reply DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoABase.cc:27]: Instead, better add it later, if it becomes needed.

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoAGPU.cc
Comment (makortel) [inline DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoAGPU.cc:31]: These would be better as local variables (because they are used in only one member function).

**Comment by arossi83** [comment]
> @sroychow @arossi83 if you haven't already, could you check that the new module works correctly
> 
> * in the `11634.501` workflow on a machine without GPUs (_e.g._ `lxplus`)
> * in the `11634.502` workflow on a machine without GPUs (_e.g._ `lxplus`)
> * in the `11634.502` workflow on a machine _with_ GPUs (_e.g._ `lxplus-gpu`)
> 
> and that the plots look OK in the three cases ?

@fwyzard the module works correctly in `11634.501` wf on lxplus and on `11634.502` wf on `lxplus-gpu` and plots are OK but it crashes on `11634.502` wf on `lxplus`. 
We use the `gpu` procModifier to select which product consume so in this "mixed" case (gpu procModifier but on a machine without a GPU) we try to access the CUDA RecHit anyway and this generates the crash. 
According to [https://github.com/cms-sw/cmssw/blob/master/HeterogeneousCore/CUDACore/README.md#analyzer-with-cuda-input](this) there is no a mechanism to avoid this actually.  if you have any suggestion please let us know.

**Comment by fwyzard** [comment]
> We use the `gpu` procModifier to select which product to consume so in this "mixed" case (gpu procModifier but on a machine without a GPU) we try to access the CUDA RecHit anyway and this generates the crash.

Mhm, I see.

The .502 workflows and the gpu modifier are supposed to work on a machine without a GPU, automatically falling back to the CPU case.

In your case I would suggest to use the `gpuValidation` modifier, as it has been done for ECAL and HCAL.

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoACPU.cc
Comment (fwyzard) [inline DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoACPU.cc:67]:
```suggestion
  desc.add<edm::InputTag>("pixelHitsSrc", edm::InputTag("siPixelRecHitsPreSplitting"));
```

File: DQM/SiPixelPhase1Heterogeneous/python/SiPixelPhase1HeterogenousDQM_FirstStep_cff.py
Comment (fwyzard) [inline DQM/SiPixelPhase1Heterogeneous/python/SiPixelPhase1HeterogenousDQM_FirstStep_cff.py:11]:
```suggestion
monitorpixelSoASourceGPU = cms.Sequence(siPixelPhase1MonitorTrackSoA * siPixelPhase1MonitorVertexSoA * siPixelPhase1MonitorRecHitsSoACPU * siPixelPhase1MonitorRecHitsSoAGPU)

from Configuration.ProcessModifiers.gpu_cff import gpuValidation
gpuValidation.toModify(siPixelPhase1MonitorRecHitsSoACPU, pixelHitsSrc = "siPixelRecHitsPreSplitting@cpu")
gpuValidation.toReplaceWith(monitorpixelSoASource, monitorpixelSoASourceGPU)
```

**Review by fwyzard — CHANGES_REQUESTED** [review]
You can check #36998 and use a similar approach:
  - define the new modifier `gpuValidationPixel` in `Configuration/ProcessModifiers/python/gpuValidationPixel_cff.py`
  - add it to the `gpuValidation` modifier in `Configuration/ProcessModifiers/python/gpuValidation_cff.py`
  - make `siPixelPhase1MonitorRecHitsSoACPU` read `siPixelRecHitsPreSplitting` by default, instead of `siPixelRecHitsPreSplitting@cpu`
  - use the `gpuValidation` modifier to read `siPixelRecHitsPreSplitting@cpu`
  - change
  ```python
monitorpixelSoAGPU = cms.Sequence(siPixelPhase1MonitorTrackSoA * siPixelPhase1MonitorVertexSoA * siPixelPhase1MonitorRecHitsSoAGPU)

from Configuration.ProcessModifiers.gpu_cff import gpu
gpu.toReplaceWith(monitorpixelSoASource, monitorpixelSoAGPU)
```
to
```python
monitorpixelSoASourceGPU = cms.Sequence(siPixelPhase1MonitorTrackSoA * siPixelPhase1MonitorVertexSoA * siPixelPhase1MonitorRecHitsSoACPU * siPixelPhase1MonitorRecHitsSoAGPU)

from Configuration.ProcessModifiers.gpuValidation_cff import gpuValidation
gpuValidation.toReplaceWith(monitorpixelSoASource, monitorpixelSoASourceGPU)
```

This should make
  - the `.501` and `.505` workflows do the validation of the CPU reconstruction
      - `siPixelPhase1MonitorRecHitsSoACPU` reads `siPixelRecHitsPreSplitting`
  - the `.502` and `.506` workflows running on CPU, do the validation of the CPU reconstruction
      - `siPixelPhase1MonitorRecHitsSoACPU` reads `siPixelRecHitsPreSplitting` that should alias `siPixelRecHitsPreSplitting@cpu`
  - the `.502` and `.506` workflows running on GPU, do the validation of the GPU reconstruction and transfer from GPU to CPU
      - `siPixelPhase1MonitorRecHitsSoACPU` reads `siPixelRecHitsPreSplitting` that should alias `siPixelRecHitsPreSplitting@cuda`
  - the `.503` and `.507` workflows (to be added in a separate PR) do the validation of both the CPU and GPU reconstruction
      - `siPixelPhase1MonitorRecHitsSoACPU` reads `siPixelRecHitsPreSplitting@cpu` explicitly
      - `siPixelPhase1MonitorRecHitsSoAGPU` reads the GPU collections

It is a somewhat different approach from the one you have used -- but it's the only way I can think of to keep the `.502` workflows working n both CPU and GPU, and is also more in line with what ECAL and HCAL have done.

**Comment by fwyzard** [comment]
> It is a somewhat different approach from the one you have used -- but it's the only way I can think of to keep the `.502` workflows working n both CPU and GPU, and is also more in line with what ECAL and HCAL have done.

_Caveat emptor_: I haven't checked if this actually works, and it's before the first coffee of the morning, so you may need to make some adjustments...

**Comment by arossi83** [comment]
@fwyzard Thanks, we will do some test! I have just a naive question:

In our case we have 3 different scenarios:
1) CPU wf : We use the siPixelPhase1MonitorRecHitsSoACPU analyzer with siPixelRecHitsPreSplitting
2) GPU wf on a machine with GPU: We use siPixelPhase1MonitorRecHitsSoAGPU with siPixelRecHitsPreSplittingCUDA
3) GPU wf on a machine without GPU : We must use siPixelPhase1MonitorRecHitsSoACPU analyzer with siPixelRecHitsPreSplitting but actually we use the same setting as the previous scenario and this generates a crash

I don't understand how we can get 3 configuration with a single modifier...

PS: if I use the same configuration of 1) in the scenario 3) everything works

**Comment by fwyzard** [comment]
> In our case we have 3 different scenarios:
> 
>     1. CPU wf : We use the siPixelPhase1MonitorRecHitsSoACPU analyzer with siPixelRecHitsPreSplitting

Yes.

>     2. GPU wf on a machine with GPU: We use siPixelPhase1MonitorRecHitsSoAGPU with siPixelRecHitsPreSplittingCUDA

My proposal is that you _do not_ use `siPixelPhase1MonitorRecHitsSoAGPU` in this case; instead, you use `siPixelPhase1MonitorRecHitsSoACPU` to read the rechits in SoA format after they have been copied form the GPU to the CPU

>     3. GPU wf on a machine without GPU : We must use siPixelPhase1MonitorRecHitsSoACPU analyzer with siPixelRecHitsPreSplitting but actually we use the same setting as the previous scenario and this generates a crash

Exactly, which is why the proposal is to change the definition of the previous scenario as well.

> I don't understand how we can get 3 configuration with a single modifier...
> 
> PS: if I use the same configuration of 1) in the scenario 3) everything works

The problem is that the modules you have written cannot be used in the workflows that have been considered so far: there is no workflow that _must_ run on GPU only, there is only a workflow that _can_ run on GPU if available, but will run on CPU if no GPU is available.

So
  - the `pixelNtupletFit` is used to run the pixel-only reconstruction on CPU (`.501`)
  - the `pixelNtupletFit` and `gpu` modifiers together alllow to run the pixel-only reconstruction on GPU, if available, or on CPU, if not (`.502`)
      - since the workflow needs to be able to run with and without GPU, it cannot include a module that _requires_ running on GPU; at some point maybe we can address this extending the functionality of the `SwitchProducer` - but today this is not possible 
      - so, the validation in this workflow is based on reading the SoA collection on the CPU; the workflow will either produce it directly on the CPU (if no GPU is available), or produce it on the GPU and copy it to the CPU (if a GPU is available)
  - the `gpuValidation` modifier makes the framework run both the CPU and GPU reconstruction, so the `pixelNtupletFit` and `gpuValidation` modifiers together can be used to request running the pixel-only reconstruction both on CPU and GPU
      -  in this case, you can use the `siPixelPhase1MonitorRecHitsSoACPU` module to monitor the SoA CPU collection, and the `siPixelPhase1MonitorRecHitsSoAGPU` to monitor the SoA GPU collection
      - you could also use two instances of the `siPixelPhase1MonitorRecHitsSoACPU` module with different configurations to monitor the SoA collection produced on the CPU (`@cpu`), and the SoA collection copied from the GPU (`@cuda`)
      - this combination of modifiers requires a new workflow, which will be `.503` (in a future PR)

**Comment by arossi83** [comment]
@fwyzard I did some test and now I'm a bit confused (as usual):

- .502 wf on NOGPU machine: if I use siPixelRecHitsPreSplitting the handle fail the .isValid() while if I use siPixelRecHitsPreSplittingSoA everything works
- .502 wf on GPU machine: if I use siPixelRecHitsPreSplitting the handle fail the .isValid() while if I use siPixelRecHitsPreSplittingSoA I get a crash with the following error
`cmsRun: /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/88f9ace6c62e1a4dd37cb8fcc59a99e4/opt/cmssw/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_3_X_2022-03-07-2300/src/RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitSoAFromLegacy.cc:220: virtual void SiPixelRecHitSoAFromLegacy::produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const: Assertion `clust.originalId() == ic' failed.` this is due to the fact that I'm opening a CUDA product with the CPUTraits

This is the reason why we decided to use directly the CUDA products via dedicated analyzer. If I'm still missing something I'm sorry but I need some further clarifications.

**Comment by fwyzard** [comment]
> if I use siPixelRecHitsPreSplittingSoA I get a crash

This is not the expected behaviour (thought it may be a known issue for a while now).
I'll try to have a look in the coming days - but it's unlikely this can make it for the coming prerelease.

**Comment by fwyzard** [comment]
Seems to be the same issue described in #35668 .

**Comment by sroychow** [comment]
@cmsbuild please test with #37193

**Comment by sroychow** [comment]
enable gpu

**Comment by mmusich** [comment]
Hi @cms-sw/dqm-l2 @cms-sw/heterogeneous-l2 is there anything holding this PR ? (the 12.3.X backport is fully signed: https://github.com/cms-sw/cmssw/pull/37242)

**Comment by arossi83** [comment]
@fwyzard if I run `.501` with the same recHits source as `.502` (i.e. `siPixelRecHitsPreSplittingSoA`) I get no recHits at all, this is why we decided to use `siPixelRecHitsPreSplitting` on CPU wf and `siPixelRecHitsPreSplittingSoA` on GPU wf

P.S.: on 12_2_X `siPixelRecHitsPreSplittingSoA` was populated also on CPU wf but from 12_3_X it is not

**Comment by fwyzard** [comment]
Mhm, that looks like something went wrong with the workflows :-/

**Comment by arossi83** [comment]
How do we proceed at this point? can this PR be merged? We can remove the modifier once (and if) wf will be fixed

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by fwyzard** [comment]
And apologies for the delay.

**Comment by jfernan2** [comment]
+1

File: DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoA.cc
Comment (makortel) [inline DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1MonitorRecHitsSoA.cc:21]: I'd suggest to remove unnecessary headers (and packages from the `BuildFile`). At least these headers
```diff
-#include "CUDADataFormats/Common/interface/Product.h"
-#include "FWCore/ServiceRegistry/interface/Service.h"
-#include "HeterogeneousCore/CUDACore/interface/ScopedContext.h"
```
and packages
```diff
-<use name="CUDADataFormats/Common"/>
-<use name="CUDADataFormats/SiPixelCluster"/>
-<use name="HeterogeneousCore/CUDACore"/>
-<use name="DataFormats/SiPixelDigi"/>
```
appear to be unused now.

**Comment by perrotta** [comment]
@sroychow could you please apply the cleaning suggested in https://github.com/cms-sw/cmssw/pull/37174#discussion_r830256751
After that, I think this PR can be merged, as well as its backport

**Comment by fwyzard** [comment]
@sroychow could you squash all commits into a single one ?

**Comment by fwyzard** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by perrotta** [comment]
+1

**Comment by fwyzard** [comment]
Yes, that seems a reasonable approach.

**Pull request merged**
